pkg/repository/prisma/buffered.go

```diff
@@ -0,0 +1,276 @@
+
+	return &IngestBuf[T, U]{
+		state:              initialized,
+		maxCapacity:        opts.maxCapacity,
+		flushPeriod:        opts.flushPeriod,
+		inputChan:          make(chan *inputWrapper[T, U], inputChannelSize),
+		lastFlush:          time.Now(),
+		internalArr:        make([]*inputWrapper[T, U], 0),
+		sizeOfData:         0,
+		maxDataSizeInQueue: opts.maxDataSizeInQueue,
+		outputFunc:         opts.outputFunc,
+		sizeFunc:           opts.sizeFunc,
+		l:                  opts.l,
+		ctx:                ctx,
+		cancel:             cancel,
+	}
+}
+
+func (b *IngestBuf[T, U]) validate() error {
+	if b.maxCapacity <= 0 {
+		return fmt.Errorf("max capacity must be greater than 0")
+	}
+	if b.flushPeriod <= 0 {
+		return fmt.Errorf("flush period must be greater than 0")
+	}
+	if b.maxDataSizeInQueue <= 0 {
+		return fmt.Errorf("max data size in queue must be greater than 0")
+	}
+	if b.outputFunc == nil {
+		return fmt.Errorf("bulk create func must be set")
+	}
+	if b.sizeFunc == nil {
+		return fmt.Errorf("size func must be set")
+	}
+	if b.l == nil {
+		return fmt.Errorf("logger must be set")
+	}
+	return nil
+}
+func (b *IngestBuf[T, U]) safeFetchSizeOfData() int {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	return b.sizeOfData
+}
+
+func (b *IngestBuf[T, U]) safeIncSizeOfData(size int) {
+	b.lock.Lock()
+	b.sizeOfData += size
+	b.lock.Unlock()
+}
+
+func (b *IngestBuf[T, U]) safeDecSizeOfData(size int) {
+	b.lock.Lock()
+	b.sizeOfData -= size
+	b.lock.Unlock()
+}
+
+func (b *IngestBuf[T, U]) safeSetLastFlush(t time.Time) {
+	b.lock.Lock()
+	b.lastFlush = t
+	b.lock.Unlock()
+}
+
+func (b *IngestBuf[T, U]) safeFetchLastFlush() time.Time {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	return b.lastFlush
+}
+
+func (b *IngestBuf[T, U]) buffWorker() {
+	for {
+		select {
+		case <-b.ctx.Done():
+			return
+		case e := <-b.inputChan:
+			b.internalArr = append(b.internalArr, e)
+			b.safeIncSizeOfData(b.calcSizeOfData([]T{e.item}))
+
+			if len(b.internalArr) >= b.maxCapacity {
+				go b.flush(b.sliceInternalArray())
+			}
+			if b.safeFetchSizeOfData() >= b.maxDataSizeInQueue {
+				go b.flush(b.sliceInternalArray())
+			}
+
+		case <-time.After(time.Until(b.safeFetchLastFlush().Add(b.flushPeriod))):
+			if len(b.internalArr) > 0 {
+				go b.flush(b.sliceInternalArray())
+			} else {
+				b.safeSetLastFlush(time.Now())
+			}
+		}
+	}
+}
+
+func (b *IngestBuf[T, U]) sliceInternalArray() (items []*inputWrapper[T, U]) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if len(b.internalArr) >= b.maxCapacity {
+		items = b.internalArr[:b.maxCapacity]
+		b.internalArr = b.internalArr[b.maxCapacity:]
+	} else {
+		items = b.internalArr
+		b.internalArr = nil
+	}
+	return items
+}
+
+type flushResponse[U any] struct {
+	result U
+	err    error
+}
+
+func (b *IngestBuf[T, U]) calcSizeOfData(items []T) int {
+	size := 0
+	for _, item := range items {
+		size += b.sizeFunc(item)
+	}
+	return size
+}
+
+func (b *IngestBuf[T, U]) flush(items []*inputWrapper[T, U]) {
```

Comment:
maybe we could use some unit tests on flush to ensure we're handling all cases of setting state post-flush. 
this looks good to me for now, but concerned this will be fragile

Reply:
Added a bunch of Unit tests.